### ISSUE TITLE
Flakey tests in FrameworkControllerTests 

### ISSUE THREAD

**Issue by ChrisMaddock** [nunit/nunit#2963]
Several tests in FrameworkControllerTests have recently failed in a flaky manner on the CI builds. We need to look into why that is, and make a fix.

The failures don't appear to be platform-specific, here are a few sample failure messages:

```
Errors, Failures and Warnings
1) Failed : NUnit.Framework.Api.FrameworkControllerTests.RunTestsAction_AfterLoad_ReturnsRunnableSuite("<filter/>")
  Expected: not null and String starting with "ID"
  But was:  "8601"
  at NUnit.Framework.Api.FrameworkControllerTests.RunTestsAction_AfterLoad_ReturnsRunnableSuite (System.String filter) [0x00000] in <1df9fb68b85648378ea6dd2e640b0ed4>:0 
```
```
1) Failed : NUnit.Framework.Api.FrameworkControllerTests.ExploreTestsAction_FilterCategory_ReturnsTests
  Expected: not null and String starting with "ID"
  But was:  "7680"
   at NUnit.Framework.Api.FrameworkControllerTests.ExploreTestsAction_FilterCategory_ReturnsTests() in /home/travis/build/nunit/nunit/src/NUnitFramework/tests/Api/FrameworkControllerTests.cs:line 316
```
```
1) Failed : NUnit.Framework.Api.FrameworkControllerTests.RunTestsAction_AfterLoad_ReturnsRunnableSuite("<filter/>")
  Expected: not null and String starting with "ID"
  But was:  "8508"
  at NUnit.Framework.Api.FrameworkControllerTests.RunTestsAction_AfterLoad_ReturnsRunnableSuite (System.String filter) [0x00000] in <9278af9c5d3e455d8858b76816878cf6>:0 
```

As always, contributions welcome!

**Comment by CharliePoole** [comment]
@ChrisMaddock Are you saying the same tests sometimes pass for the same code? I had seen this error and figured someone had messed up the setting of the prefix for the ids but if it's only failing intermittently, that explanation wouldn't apply.

**Comment by ChrisMaddock** [comment]
@CharliePoole Pass - I haven't had a chance to even look at the code, and see what these tests do! I just wanted to record the error messages as an issue.

**Comment by mikkelbu** [comment]
And again (from https://travis-ci.org/nunit/nunit/jobs/416897254)
```
1) Failed : NUnit.Framework.Api.FrameworkControllerTests.RunTestsAction_AfterLoad_ReturnsRunnableSuite("")
  Expected: not null and String starting with "ID"
  But was:  "8483"
  at NUnit.Framework.Api.FrameworkControllerTests.RunTestsAction_AfterLoad_ReturnsRunnableSuite (System.String filter) [0x00000] in <4da919d604294820b300e383ceca81a7>:0 
```

**Comment by CharliePoole** [comment]
@mikkelbu Link doesn't show that failure.

**Comment by mikkelbu** [comment]
@CharliePoole Sorry. I forgot that restarting the job cleans the log, and I restarted the job after reporting here.

**Comment by CharliePoole** [comment]
@mikkelbu I see. And it apparently succeeded when restarted.

The prefix used for all ids generated is stored in Test.IdPrefix, which is static. Maybe some other test is tromping on it.

**Comment by CharliePoole** [comment]
We might want to rethink the use of a static to store the prefix.

**Comment by mikkelbu** [comment]
I was thinking the same regarding the static, but I've only skimmed the code very quickly.

**Comment by ChrisMaddock** [comment]
Just cleared this on another PR.

A little concerned what's changed recently to cause these - they seem very frequent.

**Comment by rprouse** [comment]
I just saw this again on another PR. It is a bit worrying...

**Comment by OmicronPersei** [comment]
Wrote some instrumentation code to check for when `Test.IdPrefix` is set to something other than a string starting with "ID" and I got a callstack that looks like this

```
   at NUnit.Framework.Internal.Test.set_IdPrefix(String value)
   at NUnit.Framework.Api.SchemaTests.<>c.<TestResultSchemaMatches>b__4_0()
   at NUnit.Framework.Assert.Multiple(TestDelegate testDelegate)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at NUnit.Framework.Internal.Reflect.InvokeMethod(MethodInfo method, Object fixture, Object[] args)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.TestWorker.TestWorkerThreadProc()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```

**Comment by OmicronPersei** [comment]
Yeah so it looks like [`SchemaTests.TestResultSchemaMatches()`](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/tests/Api/SchemaTests.cs#L80)  appears to be at least one of the culprits.  Odd though that this is only causing __sporadic__ failures, not 100% of the time.

**Comment by ChrisMaddock** [comment]
That would explain why they've only started recently - great find, thanks @OmicronPersei!

**Comment by OmicronPersei** [comment]
So even if our tests that do change `Test.IdPrefix` change it back to its original state, or some valid default on test teardown, it would fix the issue, but probably not if tests are being ran in parallel?

**Comment by CharliePoole** [comment]
It has to be changed to the original state, because the runner has set it to a particular value and is possibly using that value to track tests in various assemblies. For example, the GUI maintains an index of test ids to nodes in the visual tree of tests.

But the best solution is probably that we should rethink how to provide the prefix when generating tests and stop using the static.

**Comment by ChrisMaddock** [comment]
I do think changing the prefix back to the original stage is definitely the short-term fix, for our test failures, however! 😊

**Comment by jnm2** [comment]
I'm sorry, just seeing this! I'm looking at constructor injection for test IDs.

**Comment by jnm2** [comment]
Why are things like `DefaultTestAssemblyBuilder` in `NUnit.Framework.Api`? Are users intended to be calling `new DefaultTestAssemblyBuilder()` themselves?

If so, we *will* have to rely on global (static) state to keep track of the ID prefix and next ID number.

**Comment by jnm2** [comment]
Why are things like `DefaultTestAssemblyBuilder` public in `NUnit.Framework.Api`? Are users intended to be calling `new DefaultTestAssemblyBuilder()` themselves?

If so, we *will* have to rely on global (static) state to keep track of the ID prefix and next ID number. For example, a user runs `new DefaultTestAssemblyBuilder().Build(assembly, options)` and obtains a tree of tests without specifying the prefix or starting number and yet expecting them to be used.

@CharliePoole The easy, tempting thing to do is to make `DefaultTestAssemblyBuilder` internal. I don't have enough context to know if that would be a good or bad thing to do.

**Comment by CharliePoole** [comment]
No. How to use the framework API is specified in the docs. Their are actually no classes intended to be newed by users.  Leaving things public just made it easier to use reflection.

However, IIRC Rob added some public methods for xamarin to use, so take care.

**Comment by jnm2** [comment]
Thank you, this is good to know. I don't know of a tool that lets me find all Rob's commits that affected files under the Api folder, so I'll ask him about it in the PR.

**Comment by jnm2** [comment]
Ah, this is a trickier one. `TestCaseSourceAttribute.BuildFrom` is doing `new NUnitTestCaseBuilder().BuildTestMethod`.

How can `TestCaseSourceAttribute.BuildFrom` be given access to the ID prefix and counter without making them static? Should the `TestIdProvider` I'm adding be made public and used in `ITestBuilder.BuildFrom(TestIdProvider idProvider, FixtureMethod method, Test suite)`?

**Comment by jnm2** [comment]
Ah, this is a trickier one. `TestCaseSourceAttribute.BuildFrom` is doing `new NUnitTestCaseBuilder().BuildTestMethod`.

How can `TestCaseSourceAttribute.BuildFrom` be given access to the ID prefix and counter without making them static? Should the `TestIdProvider` I'm adding be made public and added to the signature?

`ITestBuilder.BuildFrom(TestIdProvider idProvider, FixtureMethod method, Test suite)`

It would be a breaking change for anyone implementing ITestBuilder. We did recently cause a breaking change to this interface method when we replaced IMethodInfo with FixtureMethod in https://github.com/nunit/nunit/pull/2796. If we keep breaking things like this, I'm in favor of using a single `Args` parameter so we can add new parameters without breaking things.

**Comment by CharliePoole** [comment]
For the public methods, you could look at how the xamarin runner calls the framework.

**Comment by CharliePoole** [comment]
I'll need to look at the code to understand what you're asking. After I have my coffee!

**Comment by jnm2** [comment]
@CharliePoole Here's some work in progress: https://github.com/nunit/nunit/compare/test_id_injection

The problem is with all the attributes like this; there's nowhere to determine the ID which needs to be passed to `new SetUpFixture(type)` except from a method parameter or else a global static field:

![image](https://user-images.githubusercontent.com/8040367/44803091-6b12d400-ab8b-11e8-83e2-036282c44a9f.png)

**Comment by jnm2** [comment]
@CharliePoole Just force pushed. Everything's in place except XML docs and these problems which I don't know how to resolve without getting an ID provider into the BuildFrom methods of these seven methods, or giving up and going back to static field state for test IDs:

![image](https://user-images.githubusercontent.com/8040367/44804276-dd38e800-ab8e-11e8-89f6-b43997a87a81.png)

**Comment by jnm2** [comment]
Right now users who implement their own builder attributes like `ITestCaseBuilder` must create `NUnit.Framework.Internal.Test` objects. Users have two options:

- directly new up `NUnit.Framework.Internal.Test` objects
- new up `NUnit.Framework.Internal.Builders` builders and get then to create the `Internal.Test` objects

Users can't return their own implementations of ITest since the interface requires then to return `Internal.Test`.

Given all of this, what if NUnit stopped passing in `FixtureMethod method, Test suite` and instead passed in a new object like this straw man:

```cs
public sealed class TestCaseBuilderArgs
{
    public FixtureMethod Method { get; }
    public Internal.Test Suite { get; }

    // Returns a test that is initialized with an ID, the fixture method, and the suite,
    // which the user can continue initializing before returning it.
    public Internal.Test CreateTest();
}
```

This way, users who implement ITestCaseBuilder etc won't have to know or care about plumbing like TestIdProvider.

**Comment by jnm2** [comment]
Reopening to continue discussion.
